### ISSUE TITLE
Ability to remove directories

### ISSUE THREAD

**Issue by Fryguy** [rtyley/bfg-repo-cleaner#12]
I was wondering if bfg-repo-cleaner had the ability to remove entire directories of files.  I noticed the -D option, but it specifically says it doesn't work with paths, so I don't think it will work.  Would it be possible to add an option to give one or more directories?

**Comment by alexdunn** [comment]
I would also really love to see bfg support removing of specific subdirectories.  This would make it useful in my situation.

**Comment by javabrett** [comment]
+1 for this feature.  Sometimes it is prudent to prune entire paths from history.  I imagine that this is a fairly common need.  This can be achieved now, but at the moment it requires a lot of pre-BFG scripting to generate a list of objects that are in those delete-target trees but not in HEADs, then feed that to delete-by-objectId using `-bi`.  It works, but it's pretty cumbersome.

> Part of the special-sauce that makes the BFG so fast is that it is path-independent, which makes me cautious about adding path-dependent features - unless they can be added without effecting performance. Let me have a think about it, I can see it's a valid feature.

...

> ...but the code really is not setup to deal with paths, it's a non-trivial change to get it there :)

@rtyley Did you ever draw any performance-conclusions on this?  Is it just a coding-exercise without fundamental overwhelming space/time concerns?

It looks like this wants `scala-git` `Tree` to be able to maintain maps on both the blob short/relative filename, and the full path, which would require generating the two maps and the storage required for that, but doesn't otherwise seem like a big burden.

**Comment by ltrzesniewski** [comment]
Here's a workaround to remove a given directory *by path* with BFG:

```bash
git rev-list --all --objects -- path/to/the/directory/to/delete | git cat-file --batch-check='%(objectname) %(objecttype) %(rest)' | grep -Pe '^\w+ blob' | cut -d' ' -f1 > ./to-delete.txt
java -jar bfg.jar --no-blob-protection --strip-blobs-with-ids ./to-delete.txt
```

The principle is simple: create a list of object IDs to strip, and input that to BFG.

 - `git rev-list --all --objects -- path/to/the/directory/to/delete`
  This will list *all* objects in the subdirectory referenced in all commits which modify the given path. The format is `objectid filepath`.

  **You should run this command to check its output matches what you'd expect.**

 - `git cat-file --batch-check='%(objectname) %(objecttype) %(rest)'`
  This will qualify the object with its type. It will turn the previous format `objectid filepath` into `objectid type filepath`.

 - `grep -Pe '^\w+ blob'`
  This will filter out non-blob objects.

 - `cut -d' ' -f1 > ./to-delete.txt`
  This will extract the object ID and redirect the output into the `to-delete.txt` file.

 - `java -jar bfg.jar --no-blob-protection --strip-blobs-with-ids ./to-delete.txt`
  This runs BFG, giving it the list of objects to remove.

Needless to say, it's *much* faster than *git filter-branch* :smile:

**Comment by bschindler** [comment]
> Here's a workaround to remove a given directory by path with BFG:

You have to be careful with this approach. If a file has been copied from another location, this approach will also delete it in the other location as git uses the same hash for different locations. It has happened to me on trial runs a number of times. 

The best way to deal with this is to delete the directory in git first, commit and then run your script but without blob-protection. This seemed to have worked for me.

**Comment by ltrzesniewski** [comment]
@bschindler yes that's dangerous, that's what I said in bold in my comment. 
I made a PR which enables a safe method, see #166 - unfortunately the maintainer doesn't seem to care about PRs.

**Comment by rtyley** [comment]
I do, I just get through them *real* slowly

**Comment by ltrzesniewski** [comment]
@rtyley oh sorry I just noticed you started working on the project again recently

**Comment by TharosTheDragon** [comment]
I erroneously included a folder containing release builds in my git repo. Can I use BFG to undo this, removing that folder and its contents from my git history? It's just been sitting there taking up space for no reason.

**Comment by Fryguy** [comment]
@TharosTheDragon If the directory name is consistent throughout history, and doesn't conflict by name with other directories in the tree, then you could use `--delete-dirs <glob> - delete directories with the specified names`.  Note that command is based on *name*, not path, so if you have multiple directories with the same name, even at different depths, they will both be removed.

**Comment by TharosTheDragon** [comment]
What's the difference between --delete-dirs and --delete-folders?

**Comment by Fryguy** [comment]
I'm sorry...I copied the wrong string, not paying attention...should be `--delete-folders <glob>`

**Comment by TharosTheDragon** [comment]
Thanks!


________________________________

**Comment by gqy117** [comment]
Thank @ltrzesniewski for his awesome answer.
In my case, I need to delete **2 files** with full path provided.
So I tweaked @ltrzesniewski  's answer:

`git rev-list --all --objects | grep -P '^\w+ Path/to/your/file1.txt' >> ../to-delete.txt`
`git rev-list --all --objects | grep -P '^\w+ Path/to/your/file2.txt' >> ../to-delete.txt`

`java -jar bfg.jar --no-blob-protection --strip-blobs-with-ids ./to-delete.txt`
